packages/firewall/src/rate-limit.ts

```diff
@@ -0,0 +1,151 @@
+/**
+ * Check rate-limits defined through the Vercel Firewall.
+ *
+ *
+ * @param rateLimitId The ID of the rate limit to check. The same ID must be defined in the Vercel Firewall as a @vercel/firewall rule condition.
+ * @param options
+ * @returns A promise that resolves to an object with a `rateLimited` property that is `true` if the request is rate-limited, and `false` otherwise. The
+ *   `error` property is defined if the request was blocked by the firewall or the rate limit ID was not found.
+ *
+ * @example
+ * ```js
+ * import { checkRateLimit } from '@vercel/firewall';
+ *
+ * async function handler() {
+ *   const { rateLimited } = await checkRateLimit('my-rate-limit-id');
+ * }
+ * ```
+ *
+ */
+export async function checkRateLimit(
+  rateLimitId: string,
+  options?: {
+    /** The host name on which the rate limit rules are defined */
+    firewallHost?: string;
+    /** The key to use for rate-limiting. If not defined, defaults to the user's IP address. */
+    rateLimitKey?: string;
+    /** The headers for the current request. Optional if `rateLimitKey` or `request` are provided.  */
+    headers?:
+      | Headers
+      | Record<string, string>
+      | Record<string, string | string[]>;
+    /** The current request object. Optional if `rateLimitKey` or `headers` are provided */
+    request?: Request;
+  }
+): Promise<{
+  rateLimited: boolean;
+  error?: 'not-found' | 'blocked';
+}> {
+  if (process.env.NODE_ENV !== 'production' && !options?.firewallHost) {
+    console.warn(
+      'Provide the `firewallHost` option to support rate-limiting in development mode'
+    );
+    return {
+      rateLimited: false,
+    };
+  }
+
+  let requestHeaders = options?.headers || options?.request?.headers;
+  if (requestHeaders && !(requestHeaders instanceof Headers)) {
+    requestHeaders = new Headers(requestHeaders);
+  }
+  if (!requestHeaders && process.env.NEXT_RUNTIME) {
+    const { headers } = await import('next/headers');
```

Comment:
Given this will opt into dynamic rendering, we'll want to have some guidance in the docs on the best place to use this. E.g. maybe the code snippet should maybe be middleware?

Reply:
Can do. I wouldn't say that this opts you into dynamic rendering. You'd only ever call this in dynamic functions. I think `route.ts` is probably best, server actions are also great

Reply:
Changed to `route.ts` style example